Refactor BatchRepository

fourlastor opened the pull request:
This PR splits `BatchRepository` in 4 collaborators (each of which is covering one area that `BatchRepository` was covering) and renames it to `BatchFacade`:
- `BatchStatusService`
- `BatchStartingService`
- `BatchDeletionService`
- `BatchRetrievalService`

Tests have been added around the various methods of `BatchFacade` before refactoring, after the refactoring was successful, I moved the tests down one level to the collaborator and added tests for `BatchFacade` to be sure it uses its collaborators.

The only exception to this is `BatchRepository.retrieveBatchesFor` which I found **extremely** hard to test due to the intensive use of `Cursor`

blundell commented on library/build.gradle:
```diff
@@ -50,6 +50,8 @@ dependencies {
 
     testCompile 'junit:junit:4.12'
     testCompile 'org.mockito:mockito-core:1.10.19'
+    testCompile 'org.powermock:powermock-module-junit4:1.6.4'
```
:-1: powermock is not cool, I assume you've added this to either mock static methods or override final classes? A better way is to highlight where these design issues are in the code so that we can make the code better rather than using a brute force tool to get around the issue

fourlastor replied:
Discussed together and agreed that it's the least evil to be able to test it ;)

blundell replied:
:+1:

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/BatchDeletionService.java:
```diff
@@ -0,0 +1,91 @@
+class BatchDeletionService {
+
+    private final DownloadDeleter downloadDeleter;
+    private ContentResolver resolver;
```
should this be final?

otherwise split it from the others ie. whiteline space

blundell commented:
can you explain why the name change from `BatchRepository` to `BatchFacade`?  A facade is an implementation detail that you don't usually expose. The old name explained that query these methods was to do with query from a datastore.

fourlastor commented:
@blundell restored `BatchRepository` as the class name, also renamed the collaborators to be `*Repository` as we talked about yesterday :dancer: 

Also fixed `BatchDownloadsRepository` since it was breaking the tests (I introduced a regression :dancer: )

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/BatchRetrievalException.java:
```diff
@@ -0,0 +1,15 @@
+        return new BatchRetrievalException("Failed to query for batches");
+    }
+
+    static BatchRetrievalException failedQueryForBatch(long batchId) {
```
I think its a bit strange to have static factory methods on exceptions? Most people don't expect this with exceptions what about just using the constructor?

```java
public BatchRetrievalException() {
        super("Failed to query for batches");
    }

public BatchRetrievalException(long batchId) {
        super("Failed to query for batch with batchId = " + batchId);
    }
```

fourlastor replied:
Because it doesn't give it a name mainly. Also, with constructors, usually people tend to end up with the message as a parameter, leaking that information outside the class and hiding when the exception has come to a point where it makes sense to split it.

People are used to it because it's easier, imho not a good reason to keep the habit :)

blundell replied:
like I showed the message should be inside the class and so people putting it outside is not a valid reason. With the above it enforces that you cannot have the full message as a param.

Creating an exception through a factory method is very strange, when something is exceptional you should know exactly why & what exception you are using.
If you want to separate these two exceptions by name i.e. they are that different - then they should be two different classes.

fourlastor replied:
6a1dada

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/BatchStatusRepository.java:
```diff
@@ -0,0 +1,141 @@
+
+        marshallInStatuses(cursor);
+
+        cursor.close();
```
can you put this in a try finally?

fourlastor replied:
:+1: will do

fourlastor replied:
4f941f3

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/Cursors.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    private static int columnIndexFor(Cursor c, String column) {
+        return c.getColumnIndexOrThrow(column);
```
not sure it's wise to have this `OrThrow` hidden inside a method that does not explicitly say that. When I first read line 12 I was wondering why you didn't use `OrThrow` but you do just that the method is named wrong. I would inline `columnIndexFor` or rename it

fourlastor replied:
Makes sense, I'll append `orThrow` to the name.

fourlastor replied:
Actually I'll inline :dancer:

fourlastor replied:
24a972c

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/Cursors.java:
```diff
@@ -0,0 +1,27 @@
+        return c.getLong(columnIndexFor(c, column));
+    }
+
+    public static String getString(Cursor c, String column) {
```
not sure why we have these static helper methods? I would get rid of this class, next thing someone will be mocking these methods with powermock! ew

fourlastor replied:
The main idea is to simplify the use of `Cursor` which is awkward at least.

You don't need to mock this method, test wise you've the same problem (using cursor will make that part of the code hard to test), it just removes noise from the client.

fourlastor replied:
0b3e92a

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/DownloadsUriProvider.java:
```diff
@@ -59,14 +55,6 @@ private static DownloadsUriProvider newInstance() {
-     * The content URI for accessing publicly accessible downloads (i.e., it requires no
-     * permissions to access this downloaded file)
-     */
-    public Uri getPubliclyAccessibleDownloadsUri() {
```
this is removed because there is no public access now?

fourlastor replied:
It wasn't used, I suppose simply no one noticed it before.

fourlastor replied:
Oh wait, I might misunderstood the scope of this class: is this to be used by the client?

blundell replied:
yes be careful if something is public in a lin you have to really think about it. I think it was client facing :x:

fourlastor replied:
going to restore it

fourlastor replied:
a2267e3

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/Statuses.java:
```diff
@@ -0,0 +1,113 @@
+
+        int size = statusCounts.size();
+        for (int i = 0; i < size; i++) {
+            stringBuilder.append("[status: ").append(statusCounts.keyAt(i)).append(", count: ").append(statusCounts.valueAt(i)).append("]");
```
would be nicer with each append on a new line

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/Cursors.java:
```diff
@@ -0,0 +1,23 @@
+
+import android.database.Cursor;
+
+final class Cursors {
```
I don't think this class is a wise move :x: its secretly a `CursorUtil` and we don't want that. Decorate the cursor if you don't think it's methods are good enough

fourlastor commented:
Please test this again.

fourlastor commented:
@blundell what do you think of `BetterCursor` instead of `Cursors`? It's basically a wrapper with added methods. I didn't use the Android fw wrapper because it would've made the tests more complicated (not mocked exception, as usual : / )

Could be part of notils if you like the idea :tongue:

blundell commented on library/src/main/java/com/novoda/downloadmanager/lib/db/BetterCursor.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public int getInt(String columnName) {
+        return getInt(getColumnIndexOrThrow(columnName));
```
I understand your sentiment here, trying to enforce the idea of throwing all the time, which is good! However anyone who normally uses cursor will be really confused by using this method ie.. `getInt` that does `getIntOrThrow`. It seems the throw is now implicit and you should always favor explicit code. This might be a bit of over engineering, lets keep it simple? The rest of the PR is ready to ship! :)

fourlastor replied:
as in? Renaming it to `getXOrThrow`? Or just removing this class? I don't know if removing this class would really make the rest of the code cleaner :P

blundell replied:
just remove the class tbh

fourlastor commented:
@blundell  7195788

blundell commented:
LGETM :+1: see if you can find a second thumbs up now :wink:

xrigau commented on library/src/test/java/com/novoda/downloadmanager/lib/BatchStatusRepositoryTests.java:
```diff
@@ -0,0 +1,322 @@
+import static org.powermock.api.mockito.PowerMockito.whenNew;
+
+@RunWith(Enclosed.class)
+public class BatchStatusRepositoryTests {
```
syk tests :+1:

xrigau commented on library/src/main/java/com/novoda/downloadmanager/lib/Statuses.java:
```diff
@@ -0,0 +1,118 @@
+
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder("{");
```
:+1:

xrigau commented:
:+1: nice action, most changes are tests and that's goood! you shall get a :shipit:
